### ISSUE TITLE
Invoke-RestMethod Multipart/Form-Data Filename encoding with non-ascii chars

### ISSUE THREAD

**Issue by joergrenn** [PowerShell/PowerShell#19456]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

I am using the invoke-restmethod to do a Multipart/Form-Data Submission using the -Form parameter.
The form date also includes a file.
`$DocumentUploadForm = @{`
           ` file        = Get-Item $DocumentFile`
            `title       = "$DocumentTitle"`
            
When the filename contains just ascii chars the filename is encoded as expected in the request body.
However when the filename contains non-ascii chars the filename is encoded as per RFC-1342 which is unexpected.
I have compared the behavior to the CURL application and there the filename is encoded differently using unicode codepoints.

### Expected behavior

```console
Request Body for file with non-ascci chars in the filename:

...
--------------------------748339ec18b72df1
Content-Disposition: form-data; name="file"; filename="t\xcc\x88stfile.pdf"
Content-Type: application/pdf

%PDF-1.7
...

Request Body for file with only ascii chars in the filename:

...
--------------------------748339ec18b72df1
Content-Disposition: form-data; name="file"; filename="testfile.pdf"
Content-Type: application/pdf

%PDF-1.7
...
```

### Actual behavior

```console
Request Body for file with non-ascci chars in the filename:

...
--------------------------748339ec18b72df1
Content-Disposition: form-data; name="file"; filename="=?utf-8?B?S3XMiG5kaWd1bmdfTXVzdGVybWFubl9NYXguVFRBXzIwMjNfMDFfMzAucGRm?="
Content-Type: application/pdf

%PDF-1.7
...
```

### Error details

```console
There is no error in the powershell, but the server does not accept the POST request containing this filename encoding.
The server does accept a the POST request created by CURL.
The only difference is the encoding auf the filename.
Either with unicode codepoints or with RFC-1342
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.3.3
PSEdition                      Core
GitCommitId                    7.3.3
OS                             Darwin 22.4.0 Darwin Kernel Version 22.4.0: Mon Mar  6 20:59:58 PST 2023; root:xnu-8796.101.5~3/RELEASE_ARM64_T6020
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by CarloToso** [comment]
Could you please recreate this issue un a platform such as https://httpbin.org/?
And then write the steps to follow.

**Comment by joergrenn** [comment]
Hello @CarloToso ,
I don't know what I am supposed to do?
httpbin.org to me looks like a service where I can sent http requests and get the response back. 
I am afraid I am not a web developer.
What exactly would you like to look at? I have tried to describe the issue as best as I could.

**Comment by CarloToso** [comment]
@joergrenn don't worry I wasn't clear, I just wanted you to write the steps you followed to get this error, something like
`Invoke-RestMethod -Method POST "https://httpbin.org/post" -Form $DocumentUploadForm`
`curl command here`
and then the program you used to check (usually Fiddler or MITMProxy)

I looked into it and it seems it's a  .Net problem
Test:
```ps1
$contentDisposition = [System.Net.Http.Headers.ContentDispositionHeaderValue]::new("form-data")
$contentDisposition.name = "Kündigung_Mustermann_Max.TTA_2023_01_30.pdf"
[string]$contentDisposition
#-->form-data; name="=?utf-8?B?S3XMiG5kaWd1bmdfTXVzdGVybWFubl9NYXguVFRBXzIwMjNfMDFfMzAucGRm?="
```
https://github.com/PowerShell/PowerShell/blob/0ddb187e61598ff9079103b1db7e4944d422f071/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs#L1754
@iSazonov should we look into writing a workaround?

**Comment by joergrenn** [comment]
@CarloToso 
yes thats exactly what I did, I used mitmproxy to grab the request body and compare between "Invoke-RestMethod -Method POST "https://httpbin.org/post" -Form $DocumentUploadForm"
and the curl command. 
I can send you the curl command if this helps, but it seems to me you figured it out already. :)

**Comment by CarloToso** [comment]
@joergrenn I think there might also be another difference: curl -->Content-Type:application/pdf
Powershell -->Content-Type:octet/stream

**Comment by iSazonov** [comment]
It seems PowerShell doesn't FileNameStar property. 
https://github.com/PowerShell/PowerShell/blob/0ddb187e61598ff9079103b1db7e4944d422f071/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs#L1754

Related .Net code 
https://github.com/dotnet/runtime/blob/8a46f0777ef975bb3d39cfb0b477c8e5c2d02b9a/src/libraries/System.Net.Http/src/System/Net/Http/Headers/ContentDispositionHeaderValue.cs#L387-L395

So I guess a right fix is to add FileNameStar.

**Comment by CarloToso** [comment]
@iSazonov should we add FileNameStar or replace FileName with FileNameStar?

**Comment by iSazonov** [comment]
From my understanding RFC defines priority rules how process these two fields.

**Comment by daxian-dbw** [comment]
@joergrenn Now that @CarloToso has a potential fix #19467, can you please try out the private build produced from that PR and confirm whether it actually works? 

You can download the build.zip package from https://dev.azure.com/powershell/PowerShell/_build/results?buildId=127272&view=artifacts&pathAsName=false&type=publishedArtifacts:

![image](https://user-images.githubusercontent.com/127450/231237239-246a08e3-dfe9-4b8f-b3a7-e999a093afb1.png)

After you download, please remember to unblock the build.zip file before unzip it, otherwise you won't be able to start the pwsh.exe:
![image](https://user-images.githubusercontent.com/127450/231237671-ae9cbc65-6e1f-4bed-aaef-44861a27ea85.png)

**Comment by joergrenn** [comment]
Hello @daxian-dbw I am currently out of office, but next week I can certainly try it.

**Comment by joergrenn** [comment]
Hello @daxian-dbw @CarloToso I was able to test this today. 
The content-disposition in the request body now looks like this:

> Content-Disposition: form-data; name=file; 
filename="=?utf-8?B?S3XMiG5kaWd1bmdfTXVzdGVybWFubl9NYXguVFRBXzIwMjNfMDFfMzAucGRm?="; filename*=utf-8''Ku%CC%88ndigung_Mustermann_Max.TTA_2023_01_30.pdf
Content-Type: application/octet-stream
>
>%PDF-1.7
%µµµµ

I am not really familiar with the different Web Standards, but the new filename* property looks good to me.
However I am still not sure about  the filename property.
According to rfcs I looked at, the filename property is supposed to be encoded in ISO-8859-1 (extended ASCII latin1), but I have also seen just US-ASCII supported.

The filename property technically is ASCII, but further encoded using rfc1342.
I don't know if applications expect this or test for this.
My gut feeling says that in the filename property the filename should not be further encoded, but all non ASCII chars should be omitted or replaced.
In this case the filename would look like this:
filename="Kundigung_Mustermann_Max.TTA_2023_01_30.pdf";
filename*=utf-8''Ku%CC%88ndigung_Mustermann_Max.TTA_2023_01_30.pdf"

But I don't know if this is specified in rfcs somewhere what the defined bahaviour would be.

**Comment by joergrenn** [comment]
@CarloToso No, I am afraid it still does not work in my use case. The webservice I want to use does not support the filename* property, only filename at the moment.
And the service does not expect the RFC-1342 encoding of the filename property. :(
This is why I asked if it makes sense to omit the RFC-1342 encoding in the filename property.
I don't know which implementation adheres to the standards.
I have also asked the developers of the webservice to support filename*.

**Comment by iSazonov** [comment]
I set Fixed in #19467. If it is not true, please open new issue.
